lib/maths/CTreeShapFeatureImportance.cc

```diff
@@ -157,25 +158,34 @@ void CTreeShapFeatureImportance::shapRecursive(const TTree& tree,
         }
         double incomingFractionZero{1.0};
         double incomingFractionOne{1.0};
-        int splitFeature = static_cast<int>(tree[nodeIndex].splitFeature());
-        auto it = std::find(splitPath.s_FeatureIndex.begin(),
-                            splitPath.s_FeatureIndex.end(), splitFeature);
-        if (it != splitPath.s_FeatureIndex.end()) {
+        int splitFeature{static_cast<int>(tree[nodeIndex].splitFeature())};
+        auto featureIndexEnd{(splitPath->s_FeatureIndex.begin() + splitPath->nextIndex())};
+        auto it = std::find(splitPath->s_FeatureIndex.begin(), featureIndexEnd, splitFeature);
+        if (it != featureIndexEnd) {
+            // since we path splitPath by reference, we need ot backup the object before unwinding it
+            backupPath = std::make_unique<SPath>(*splitPath);
             auto pathIndex = static_cast<std::size_t>(
-                std::distance(splitPath.s_FeatureIndex.begin(), it));
-            incomingFractionZero = splitPath.fractionZeros(pathIndex);
-            incomingFractionOne = splitPath.fractionOnes(pathIndex);
-            CTreeShapFeatureImportance::unwindPath(splitPath, pathIndex);
+                std::distance(splitPath->s_FeatureIndex.begin(), it));
+            incomingFractionZero = splitPath->fractionZeros(pathIndex);
+            incomingFractionOne = splitPath->fractionOnes(pathIndex);
+            CTreeShapFeatureImportance::unwindPath(*splitPath, pathIndex);
         }
 
         double hotFractionZero = samplesPerNode[hotIndex] / samplesPerNode[nodeIndex];
         double coldFractionZero = samplesPerNode[coldIndex] / samplesPerNode[nodeIndex];
+        std::size_t nextIndex = splitPath->nextIndex();
         this->shapRecursive(tree, samplesPerNode, encoder, encodedRow, splitPath,
                             hotIndex, incomingFractionZero * hotFractionZero,
                             incomingFractionOne, splitFeature, offset, row);
+        this->unwindPath(*splitPath, nextIndex);
         this->shapRecursive(tree, samplesPerNode, encoder, encodedRow, splitPath,
                             coldIndex, incomingFractionZero * coldFractionZero,
                             0.0, splitFeature, offset, row);
+        this->unwindPath(*splitPath, nextIndex);
+        if (backupPath) {
+            // now we swap to restore the data before unwinding
+            backupPath.swap(splitPath);
```

Comment:
If splitPath is by reference then this would just be:
```
splitPath = std::move(backupPath);
```

Reply:
Done